### PR TITLE
Qt/initial unlock wallet flow

### PR DESCRIPTION
**Pull request by ferrerrajc**

### CODE REVIEW

File: src/qt/forms/enterunlockcode.ui [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+             <property name="leftMargin">
+              <number>0</number>
+             </property>
+             <property name="topMargin">
+              <number>0</number>
+             </property>
+             <item>
+              <widget class="QLabel" name="infoText">
+               <property name="text">
+                <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Merit is an invite-only network. Enter your unlock code below to unlock your wallet and start using Merit. &lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
```
Comment (ferrerrajc) [inline src/qt/forms/enterunlockcode.ui]: This could probably use some better wording

File: src/qt/enterunlockcode.cpp [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    animation->start(QAbstractAnimation::DeleteWhenStopped);
+    layerIsVisible = !hide;
+}
+
+void EnterUnlockCode::setModel(WalletModel *model)
+{
+    this->walletModel = model;
+}
+
+void EnterUnlockCode::closeClicked()
```
Comment (mempko) [inline src/qt/enterunlockcode.cpp]: remove this

File: src/qt/enterunlockcode.cpp [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+void EnterUnlockCode::submitButtonClicked()
+{
+    CMeritAddress parentAddress
+    {
+        ui->unlockCodeTextInput->toPlainText().toStdString()
+    };
+
+    if (!parentAddress.IsValid()) {
+        // TODO: handle it
```
Comment (mempko) [inline src/qt/enterunlockcode.cpp]: Use standard QT error dialog here. Example int ret = QMessageBox::warning(this, tr("My Application"),
                               tr("The document has been modified.\n"
                                  "Do you want to save your changes?"),
                               QMessageBox::Save | QMessageBox::Discard
                               | QMessageBox::Cancel,
                               QMessageBox::Save)

File: src/qt/enterunlockcode.h [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+Q_SIGNALS:
+    void walletReferred();
+
+public Q_SLOTS:
+    void submitButtonClicked();
+    void toggleVisibility();
+    void setModel(WalletModel *model);
+    // will show or hide the modal layer
+    void showHide(bool hide = false, bool userRequested = false);
+    void closeClicked();
```
Comment (mempko) [inline src/qt/enterunlockcode.h]: remove closeClicked

File: src/qt/enterunlockcode.h [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+public:
+    explicit EnterUnlockCode(QWidget *parent);
+    ~EnterUnlockCode();
+
+Q_SIGNALS:
+    void walletReferred();
+
+public Q_SLOTS:
+    void submitButtonClicked();
+    void toggleVisibility();
```
Comment (mempko) [inline src/qt/enterunlockcode.h]: remove toggleVisibility

File: src/qt/enterunlockcode.cpp [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    else if (ev->type() == QEvent::ParentChange) {
+        if (parent()) {
+            parent()->installEventFilter(this);
+            raise();
+        }
+    }
+    return QWidget::event(ev);
+}
+
+void EnterUnlockCode::toggleVisibility()
```
Comment (mempko) [inline src/qt/enterunlockcode.cpp]: remove toggleVisibility

File: src/qt/meritgui.cpp [hunk 7]
```diff
@@ -120,6 +121,7 @@ MeritGUI::MeritGUI(const PlatformStyle *_platformStyle, const NetworkStyle *netw
     rpcConsole(0),
     helpMessageDialog(0),
     modalOverlay(0),
+    enterUnlockCode(0),
```
Comment (mempko) [inline src/qt/meritgui.cpp]: use nullptr

File: src/qt/walletmodel.cpp [hunk 8]
```diff
@@ -413,6 +413,16 @@ WalletModel::EncryptionStatus WalletModel::getEncryptionStatus() const
     }
 }
 
+bool WalletModel::IsReferred() const
+{
+    return wallet->IsReferred();
+}
+
+referral::ReferralRef WalletModel::Unlock(const referral::Address& parentAddress)
+{
```
Comment (mempko) [inline src/qt/walletmodel.cpp]: assert(wallet)

File: src/qt/enterunlockcode.cpp [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    if (!parentAddress.IsValid()) {
+        // TODO: handle it
+    }
+
+    auto parentAddressUint160 = parentAddress.GetUint160();
+    if(parentAddressUint160) {
+        referral::ReferralRef referral = walletModel->Unlock(*parentAddressUint160);
+        if(referral) {
+            Q_EMIT walletReferred();
+        }
```
Comment (mempko) [inline src/qt/enterunlockcode.cpp]: if !referral then maybe show a dialog saying there was a problem unlocking.

**Pull request merged**